### PR TITLE
query endpoint / zapi query command

### PR DESCRIPTION
**Pull request by mattnibs**
Add new zed lake query endpoint and zapi command that consumes it.

Changes to query endpoint vs deprecated search endpoint:
- Query endpoint accepts zed queries as strings NOT generated ASTs.
- Adjustments to the control message protocol (supported for zng and
  zjson responses):
  - The QueryChannelSet control message indicates the current channel
    for the stream- all subsequent records will be part of the set
    channel until another QueryChannelSet message is received. In the
    old search endpoint, multi channel searches were not supported. They
    are now supported in the new query endpoint.
  - zjson "records" or "objects" are no longer batched. Each record is
    wrapped in an "Object" response.

Part of #2651

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -147,3 +147,30 @@ type IndexPostRequest struct {
 	Patterns []string `json:"patterns"`
 	Zed      string   `json:"zed,omitempty"`
 }
+
+type QueryRequest struct {
+	Query    []string `json:"query"`
+	Includes []string `json:"string"`
```
Comment (nwt) [inline api/api.go:153]:
```suggestion
	Query string `json:"query"`
```

File: service/handlers.go [hunk 2]
```diff
@@ -27,6 +30,39 @@ import (
 	"go.uber.org/zap"
 )
 
+func handleQuery(c *Core, w *ResponseWriter, r *Request) {
+	var req api.QueryRequest
+	if !r.Unmarshal(w, &req) {
+		return
+	}
+	zquery, err := query.ParseSources(req.Query, req.Includes)
```
Comment (nwt) [inline service/handlers.go:38]: This reads from the file system, so it needs to happen in the client.

Comment (mattnibs) [reply service/handlers.go:38]: Cool

File: api/queryio/writer.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
+package queryio
```
Comment (nwt) [inline api/queryio/writer.go:1]: Name this file `driver.go`?

File: api/queryio/client.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	case *api.QueryChannelSet:
+		err = r.flush()
+		r.cid = ctrl.ChannelID
+	case *api.QueryChannelEnd:
+		err = r.driver.ChannelEnd(ctrl.ChannelID)
+	case *api.QueryStats:
+		r.stats.BytesRead = ctrl.ScannerStats.BytesRead
+		r.stats.BytesMatched = ctrl.ScannerStats.BytesMatched
+		r.stats.RecordsRead = ctrl.ScannerStats.RecordsRead
+		r.stats.RecordsMatched = ctrl.ScannerStats.RecordsMatched
```
Comment (nwt) [inline api/queryio/client.go:83]:
```suggestion
		r.stats = zbuf.ScannerStats(ctrl.ScannerStats)
```

File: api/queryio/client.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+func (r *runner) flush() error {
+	if len(r.recs) > 0 {
+		recs := make([]*zng.Record, len(r.recs))
+		copy(recs, r.recs)
+		r.recs = r.recs[:0]
+		return r.driver.Write(r.cid, zbuf.Array(recs))
+	}
+	return nil
+}
```
Comment (nwt) [inline api/queryio/client.go:69]: Maybe simplify by getting rid of `flush` and always handing single-record batches to the driver? We can always get more complicated down the road if the extra allocations become a bottleneck.
```suggestion
func (r *runner) Write(rec *zng.Record) error {
	return r.driver.Write(r.cid, &zbuf.Array{rec})
}
```

**Review by nwt — APPROVED** [review]

File: service/ztests/service.sh [hunk 6]
```diff
@@ -22,7 +22,7 @@ mkdir -p $lakeroot
 
 portdir=$(mktemp -d)
 
-zed lake serve -l=localhost:0 -R $lakeroot -portfile=$portdir/lake -log.level=debug $LAKE_EXTRA_FLAGS &> lake.log &
+zed lake serve -l=localhost:0 -R $lakeroot -portfile=$portdir/lake -log.level=warn $LAKE_EXTRA_FLAGS &> lake.log &
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply service/ztests/service.sh:25]: Yeah but should be a different pr.

File: api/queryio/client.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		}
+		if ctrl != nil {
+			run.handleCtrl(ctrl)
+			continue
+		}
+		if rec != nil {
+			run.Write(rec)
+			continue
+		}
+		return run.stats, run.flush()
```
Comment (nwt) [inline api/queryio/client.go:41]:
```suggestion
		if rec == nil {
			return run.stats, nil
		}
		if err := run.Write(rec); err != nil {
			return run.stats, err
		}
```

File: api/queryio/client.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+func (r *runner) Write(rec *zng.Record) error {
+	return r.driver.Write(r.cid, &zbuf.Array{rec})
+}
+
+func (r *runner) handleCtrl(ctrl interface{}) error {
+	var err error
+	switch ctrl := ctrl.(type) {
+	case *api.QueryChannelSet:
+		err = r.flush()
```
Comment (nwt) [inline api/queryio/client.go:61]:
```suggestion
```

**Pull request merged**